### PR TITLE
Split `microovn.central` service.

### PR DESCRIPTION
**Pull request by mkalcok**
Previously `microovn.central` started all three central services (ovn-nb, ovn-sb and northd). Now these daemons are split into their own services for more fine-grained controll over them.

The `microovn.central` will stay for now and act as an umbrella service that can trigger start of all three services. However the usage is discouraged as it will be removed. It's also not capable of being used as shortcut for stopping these services.

Additional consequence of splitting of the services is that we can now scale the cluster without restarting OVN NB/SB services as only the northd daemon needs restart.

### CODE REVIEW

File: snapcraft/ovn-central.env [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+export OVN_SYSCONFDIR="${SNAP}/etc"
+export OVN_PKIDIR="${SNAP_COMMON}/data/pki"
+
+# Disable some commands
+mkdir -p "${OVN_RUNDIR}/bin/"
+for i in install plymouth sudo systemctl; do
+    [ -e "${OVN_RUNDIR}/bin/${i}" ] && continue
+
+    ln -s "/bin/true" "${OVN_RUNDIR}/bin/${i}"
+done
```
Comment (fnordahl) [inline snapcraft/ovn-central.env:20]: While I understand how these ended up in this file, it's probably not the correct place to host them after the split.

Having three services starting independently attempting to create this might lead to a race conditions and spurious errors. (And having an sourceable "environment" file performing actions behind your back comes as an element of surprise.)

Would it make sense to move this to either the install hook or as part of the startup of the `daemon` service?

File: docs/reference/services.rst [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    snap services microovn
+
+microovn.central
+----------------
+
+.. warning::
+
+    Deprecated: This service will be removed in future release
+
+This is a shortcut service. Starting this service will start and enable
```
Comment (fnordahl) [inline docs/reference/services.rst:20]: s/shortcut/transitional

File: docs/reference/services.rst [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+``microovn.ovn-nb``, ``microovn.ovn-sb`` and ``microovn.ovn-northd``. However
+this service , due to the ``snap`` limitations, is not capable of stopping
+child services and its usage is generally discouraged.
```
Comment (fnordahl) [inline docs/reference/services.rst:23]: I'd rather highlight the reasons for why managing these services separately is desirable rather than bashing snaps here

File: snapcraft/commands/central.start [hunk 4]
```diff
@@ -1,67 +1,15 @@
...
-
-# Start SouthBound OVN DB
-"${SNAP}/share/ovn/scripts/ovn-ctl" run_sb_ovsdb ${OVN_ARGS} &
-
-# Start NorthBOund daemon
-"${SNAP}/share/ovn/scripts/ovn-ctl" start_northd ${OVN_ARGS} \
-    --ovn-manage-ovsdb=no --no-monitor
-
-sleep infinity
+echo "This service can be used as a shortcut to start ovn-nb, ovn-sb and \
```
Comment (fnordahl) [inline snapcraft/commands/central.start:3]: s/shortcut/transitional

**Review by tomponline — COMMENTED** [review]
I'm looking forward to this landing :)

File: microovn/ovn/bootstrap.go [hunk 5]
```diff
@@ -68,9 +68,9 @@ func Bootstrap(s *state.State) error {
 	// all OVS-based programs ability to specify active or passive (listen)
 	// connection types.
 	err = GenerateNewServiceCertificate(s, "client", CertificateTypeServer)
-        if err != nil {
-                return fmt.Errorf("failed to generate TLS certificate for client: %s", err)
-        }
+	if err != nil {
+		return fmt.Errorf("failed to generate TLS certificate for client: %s", err)
+	}
```
[the remark this answers is not in the document]
Comment (mkalcok) [reply microovn/ovn/bootstrap.go:73]: MicroOVN seems to be using tabs for indentation and this piece of code used spaces.This was an auto-formater change and I'd leave it in for consistency.

File: microovn/ovn/bootstrap.go [hunk 6]
```diff
@@ -93,9 +93,19 @@ func Bootstrap(s *state.State) error {
 	}
 
 	// Enable OVN central.
-	err = snapStart("central", true)
+	err = snapStart("ovn-nb", true)
```
[the remark this answers is not in the document]
Comment (mkalcok) [reply microovn/ovn/bootstrap.go:96]: Ideally users should not interact with these services at all...but I see your point and I don't see any downside to this change, aside from bit of verbosity

File: microovn/ovn/bootstrap.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (fnordahl) [reply microovn/ovn/bootstrap.go:73]: Cleaning up inconsistencies is always a nice thing, but perhaps they belong in a separate commit?

File: snapcraft/ovn-central.env (hunk 1 above)
Comment (fnordahl) [reply snapcraft/ovn-central.env:20]: I'd rather not have them in a sourceable `.env` file.

Speaking of who needs this, I honestly do not think this is related to any of the central services at all, I do not find any evidence of the `ovn-ctl`, `ovs-ctl` or `ovs-lib` scripts making calls to these binaries. So I wonder what was the reason for the original introduction of this, habits of the original author?

File: docs/reference/services.rst (hunk 3 above)
Comment (fnordahl) [reply docs/reference/services.rst:23]: Sure, I'd angle it with a message telling the user to use the other services to stop things instead then. Together with the transitional label it will make it absolutely clear to the user that they should stop whatever it is they think they should be doing and follow the new workflow

File: docs/reference/services.rst [hunk 7]
```diff
@@ -0,0 +1,63 @@
+=================
+MicroOVN Services
```
Comment (pmatulis) [inline docs/reference/services.rst:2]: s/Services/services

File: docs/reference/index.rst [hunk 8]
```diff
@@ -0,0 +1,11 @@
+==========
+References
+==========
+
+These references describe concepts unique to MicroOVN. They do not aim to
+describe general OVN/OVS concepts
+
```
Comment (pmatulis) [inline docs/reference/index.rst:7]: MicroOVN reference material is specific to the MicroOVN project. It does not cover upstream OVN/OVS topics.

File: docs/reference/services.rst [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+=================
+
+This document provides brief explanation of services provided by MicroOVN. They
+are generally not meant for direct user interaction and should be used only
+if necessary. List and status of these services can be displayed by running:
+
```
Comment (pmatulis) [inline docs/reference/services.rst:8]: This page presents a list of all MicroOVN services. The descriptions presented here is for reference only - the user is not expected to interact directly with these services.

The status of all services is displayed by running:

File: docs/reference/services.rst [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+.. code-block:: none
+
+    snap services microovn
```
Comment (pmatulis) [inline docs/reference/services.rst:11]: Align the first letter of the command ('s') directly under the 'c' (of code-block). This should be done everywhere in the document, even for 'note' or 'warning', etc.

File: docs/reference/services.rst [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+microovn.central
+----------------
+
+.. warning::
+
+    Deprecated: This service will be removed in future release
```
Comment (pmatulis) [inline docs/reference/services.rst:18]: The ``microovn.central`` service is deprecated and will be removed in a future release.

File: docs/reference/services.rst (hunk 2 above)
Comment (pmatulis) [inline docs/reference/services.rst:20]: ... and enable multiple services:

* ``microovn.ovn-nb``
* ``microovn.ovn-sb``
* ``microovn.ovn-northd``

However, ...

File: docs/reference/services.rst [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+This document provides brief explanation of services provided by MicroOVN. They
+are generally not meant for direct user interaction and should be used only
+if necessary. List and status of these services can be displayed by running:
+
+.. code-block:: none
+
+    snap services microovn
+
+microovn.central
```
Comment (pmatulis) [inline docs/reference/services.rst:13]: Let's put this in monospace (double backticks) - same for the rest of the headers with service names.

File: docs/reference/services.rst [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+
+This is a shortcut service. Starting this service will start and enable
+``microovn.ovn-nb``, ``microovn.ovn-sb`` and ``microovn.ovn-northd``. However
+this service , due to the ``snap`` limitations, is not capable of stopping
+child services and its usage is generally discouraged.
+
+microovn.chassis
+----------------
+
+This service maps directly to ``ovn-controller`` daemon.
```
Comment (pmatulis) [inline docs/reference/services.rst:28]: ... the ``ovn-controller``

File: docs/reference/services.rst [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+
+microovn.daemon
+---------------
+
+Main MicroOVN service/process that manages rest of the processes. It also
```
Comment (pmatulis) [inline docs/reference/services.rst:33]: The main ...

... manages all the other processes.

File: docs/reference/services.rst [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+handles communication with other MicroOVN cluster members and provides an API
+for ``microovn`` client command.
```
Comment (pmatulis) [inline docs/reference/services.rst:35]: ... for the ``microovn`` ...

File: docs/reference/services.rst [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+
+microovn.ovn-nb
+---------------
+
+This service maps directly to ``OVN Northbound`` database/service.
```
Comment (pmatulis) [inline docs/reference/services.rst:40]: ... directly to the ...

File: docs/reference/services.rst [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+
+microovn.ovn-northd
+-------------------
+
+This service maps directly to ``ovn-northd`` daemon.
```
Comment (pmatulis) [inline docs/reference/services.rst:45]: ... directly to the ...

File: docs/reference/services.rst [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+
+microovn.ovn-sb
+---------------
+
+This service maps directly to ``OVN Southbound`` database/service.
```
Comment (pmatulis) [inline docs/reference/services.rst:50]: ... directly to the ...

File: docs/reference/services.rst [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+
+microovn.refresh-expiring-certs
+-------------------------------
+
+This service is a recurring process that runs once a day between ``02:00`` and
+``02:30``. It triggers TLS certification reissue for certificates that are
+nearing the expiration. For more information see ``Certificate lifecycle``
```
Comment (pmatulis) [inline docs/reference/services.rst:57]: ... information see the ...

File: docs/reference/services.rst [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+section in :doc:`/how-to/tls`
```
Comment (pmatulis) [inline docs/reference/services.rst:58]: Link directly to that section. See:

https://docs.openstack.org/charm-guide/latest/community/doc-contrib/rst.html#linking-to-a-location-within-a-page-in-the-doc-set

**Review by pmatulis — CHANGES_REQUESTED** [review]
Thanks for this. See change requests.

I would normally expect to see a significant doc change like this as a separate PR with a separate Jira card.

File: snapcraft/ovn-central.env (hunk 1 above)
Comment (mkalcok) [reply snapcraft/ovn-central.env:20]: I'd be in favor of removing this then. The symlinks do not even have effect without the very specific `{$OVN_RUNDIR}/bin/` added to path, and that happens only here.

File: docs/reference/index.rst [hunk 21]
```diff
@@ -0,0 +1,11 @@
+==========
+References
```
[the remark this answers is not in the document]
Comment (mkalcok) [reply docs/reference/index.rst:2]: I followed naming convention in other documentation modes. Both `How-to guides` and `Tutorials` use plural.

**Comment by mkalcok** [comment]
Thank you for extensive review @fnordahl @pmatulis . I believe I addressed all the comments.

> I would normally expect to see a significant doc change like this as a separate PR

The need for this documentation was created directly by code changes in this PR. We split the `microovn.central` service but retained it (for some underlying technical reasons) which, imo, created a bit of an ambiguity. That's why we decided to include also this documentation piece that would explain the purpose of each service exposed by the MicroOVN snap.

File: docs/reference/services.rst [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+
+   The ``microovn.central`` service is deprecated and will be removed in a
+   future release.
+
+This is a transitional service. Starting this service will start and enable
+multiple services:
+
+* ``microovn.ovn-ovsdb-server-nb``
+* ``microovn.ovn-ovsdb-server-sb``
+* ``microovn.ovn-northd``.
```
Comment (pmatulis) [inline docs/reference/services.rst:28]: extra dot

File: docs/reference/services.rst [hunk 23]
```diff
@@ -0,0 +1,71 @@
+=================
+MicroOVN services
+=================
+
+This page presents a list of all MicroOVN services. The descriptions presented
```
Comment (pmatulis) [inline docs/reference/services.rst:5]: apologies, my fault here:

s/The descriptions presented here is/Their descriptions are/

**Review by pmatulis — CHANGES_REQUESTED** [review]
Sorry, a few last things

File: snapcraft/commands/central.start [hunk 24]
```diff
@@ -1,67 +1,15 @@
...
-export OVN_PKGDATADIR="${SNAP}/share/ovn"
-export OVN_SYSCONFDIR="${SNAP}/etc"
-export OVN_PKIDIR="${SNAP_COMMON}/data/pki"
-
-# Disable some commands
-mkdir -p "${OVN_RUNDIR}/bin/"
-for i in install plymouth sudo systemctl; do
-    [ -e "${OVN_RUNDIR}/bin/${i}" ] && continue
-
-    ln -s "/bin/true" "${OVN_RUNDIR}/bin/${i}"
```
Comment (fnordahl) [inline snapcraft/commands/central.start:20]: Good riddance!

However, I see these are also in `snapcraft/commands/chassis.start` and `snapcraft/commands/switch.start`.

Would it be more consistent to drop all of these simultaneously in a separate parent commit?


For the record, there are now some additional `DENIED` messages logged, but they do not affect operation.
```
[  101.595042] audit: type=1400 audit(1697770466.057:241): apparmor="DENIED" operation="exec" profile="snap.microovn.ovn-ovsdb-server-nb" name="/usr/bin/systemctl" pid=2804 comm="ovn-ctl" requested_mask="x" denied_mask="x" fsuid=0 ouid=0
[  101.595086] audit: type=1400 audit(1697770466.057:242): apparmor="DENIED" operation="exec" profile="snap.microovn.ovn-ovsdb-server-nb" name="/usr/bin/systemctl" pid=2804 comm="ovn-ctl" requested_mask="x" denied_mask="x" fsuid=0 ouid=0
[  101.596966] audit: type=1400 audit(1697770466.061:243): apparmor="DENIED" operation="exec" profile="snap.microovn.ovn-ovsdb-server-nb" name="/usr/bin/plymouth" pid=2806 comm="ovn-ctl" requested_mask="x" denied_mask="x" fsuid=0 ouid=0
[  101.597023] audit: type=1400 audit(1697770466.061:244): apparmor="DENIED" operation="exec" profile="snap.microovn.ovn-ovsdb-server-nb" name="/usr/bin/plymouth" pid=2806 comm="ovn-ctl" requested_mask="x" denied_mask="x" fsuid=0 ouid=0
```

Even without this change there are still other unhandled `DENIED` rules logged, such as:
```
[  101.599507] audit: type=1400 audit(1697770466.065:245): apparmor="DENIED" operation="capable" profile="snap.microovn.switch" pid=2787 comm="ovsdb-server" capability=38  capname="perfmon"
```

So I still think we should still drop these hacks and figure out what indirect dependencies of the OVS/OVN ctl scripts end up calling the offending binaries and provide patches upstream at some point in the future.

File: docs/how-to/tls.rst [hunk 25]
```diff
@@ -112,6 +112,8 @@ member.
    member. Any failure will result in subsequent communication errors for that
    service within the cluster.
 
+.. _certificates_lifecycle:
```
Comment (fnordahl) [inline docs/how-to/tls.rst:115]: Move the doc changes to a separate commit? Some of these changes will surely cause conflicts on backport, and it would be very nice to not have to deal with conflicts on the actual code changes.

**Review by fnordahl — CHANGES_REQUESTED** [review]
I took this for a spin and tested both refresh, fresh install and post configuration reboots, looks great.

The commit messages may need updating to reflect changes during review.

I also have a couple of comments in-line that request splitting a couple of things into separate commits.

File: snapcraft/commands/central.start (hunk 24 above)
Comment (mkalcok) [reply snapcraft/commands/central.start:20]: I agree. I'll drop those instances too. The commands do not work anyway, it's better to have the failures logged rather than silently failing.

File: docs/how-to/tls.rst (hunk 25 above)
Comment (mkalcok) [reply docs/how-to/tls.rst:115]: Yeah. I didn't slice this PR into commits very well.

**Comment by mkalcok** [comment]
Thanks again for the review. I incorporated your comments and I separated the docs and the command masking removal into separate commits.

**Review by fnordahl — APPROVED** [review]
LGTM, thanks!
